### PR TITLE
support config for private image hub

### PR DESCRIPTION
**Pull request by abikouo**
- new custom configuration for virt lightning on {storage_dir}/config/conf.yaml, this should allow to define private image hub using key ``images_url``
- now when provisioning a vm using ``up`` or ``start``, the missing image will be automatically fetched from the list of images url (private one + default one)
- add new shell fonctionalities to read/update custom configuration

### CODE REVIEW

File: virt_lightning/api.py [hunk 1]
```diff
@@ -408,16 +414,36 @@ def storage_dir(configuration, **kwargs):
...
     """
-    Retrieve a VM image from Internet.
+    Return the location of the VM image storage directory.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_storage_pool(configuration.storage_pool)
+    conf = hv.get_vl_configuration()
+    if kwargs.get('output') == "json" : 
+        import json
```
Comment (zloidemon) [inline virt_lightning/api.py:426]: Why do you use import not from head of file?

File: virt_lightning/api.py [hunk 2]
```diff
@@ -442,31 +467,60 @@ def redirect_request(self, req, fp, code, msg, hdrs, newurl):
     logger.debug("Date: %s", last_modified)
     size = r.headers["Content-Length"]
     logger.debug("Size: %s", size)
-    lenght = int(r.headers["Content-Length"])
+    length = int(r.headers["Content-Length"])
```
Comment (zloidemon) [inline virt_lightning/api.py:470]: You can use `int(size)` to avoid duplication access to headers.

File: virt_lightning/api.py [hunk 3]
```diff
@@ -442,31 +467,60 @@ def redirect_request(self, req, fp, code, msg, hdrs, newurl):
...
+    if custom_images_url is not None:
+        for images_url in custom_images_url :
+            try:
+                fetch_from_url(progress_callback=progress_callback,
+                    storage_dir=hv.get_storage_dir(),
+                    url=images_url,
+                    **kwargs)
+                image_found = True
+                break
+            except ImageNotFoundUpstream:
```
Comment (zloidemon) [inline virt_lightning/api.py:516]: Add more logging here to avoid removing in future.

**Review by zloidemon — CHANGES_REQUESTED** [review]

File: virt_lightning/api.py [hunk 4]
```diff
@@ -408,16 +414,36 @@ def storage_dir(configuration, **kwargs):
     hv.init_storage_pool(configuration.storage_pool)
     return hv.get_storage_dir()
 
-
-def fetch(configuration, progress_callback=None, **kwargs):
+def custom_config_view(configuration, **kwargs):
     """
-    Retrieve a VM image from Internet.
```
Comment (goneri) [inline virt_lightning/api.py:414]: There is already a `virt_lightning.configuration` module which uses `~/.config/virt-lightning/config.ini` as configuration file. I think we can just add an extra key there pointing on the 2nd repository (or better a list of extra repositories).

File: virt_lightning/api.py (hunk 2 above)
Comment (goneri) [reply virt_lightning/api.py:470]: He's actually just fixed my typo :-D.

File: virt_lightning/shell.py [hunk 5]
```diff
@@ -230,6 +229,35 @@ def list_from_yaml_file(value):
     )
     fetch_parser.add_argument("distro", help="Name of the VM image", type=str)
 
+    # custom configuration parser
```
Comment (goneri) [inline virt_lightning/shell.py:232]: I'm not sure we need this to resolve the problem if we document how to edit the configuration. This being said the idea I like `--output` parameter, can we take the time to speak about first?

File: virt_lightning/configuration.py [hunk 6]
```diff
@@ -61,7 +59,11 @@ def __init__(self):
             self.load_file(DEFAULT_CONFIGFILE.expanduser())
 
     def __get(self, key):
-        return self.data.get("main", key)
+        try:
```
Comment (goneri) [inline virt_lightning/configuration.py:62]: I'm not totally sure we need this construction. I would put everything in `main`. Overall I feel like the use of the INI format is unfortunate :-).

**Comment by goneri** [comment]
PR works great! Thank @abikouo. We can merge it as soon as the code is properly formatted.

File: virt_lightning/api.py [hunk 7]
```diff
@@ -447,31 +448,60 @@ def redirect_request(self, req, fp, code, msg, hdrs, newurl):
...
+def fetch(configuration=None, progress_callback=None,hv=None,**kwargs):
+    """
+    Retrieve a VM image from Internet.
+    """
+    if hv is None :
+        conn = libvirt.open(configuration.libvirt_uri)
+        hv = vl.LibvirtHypervisor(conn)
+        hv.init_storage_pool(configuration.storage_pool)
+    
+    configuration = Configuration()
```
Comment (goneri) [inline virt_lightning/api.py:486]: This overwrite the `configuration` parameter.

**Review by goneri — CHANGES_REQUESTED** [review]

File: virt_lightning/api.py [hunk 8]
```diff
@@ -447,31 +449,67 @@ def redirect_request(self, req, fp, code, msg, hdrs, newurl):
...
+def fetch(configuration=None, progress_callback=None, hv=None, **kwargs):
+    """
+    Retrieve a VM image from Internet.
+    """
+    if hv is None:
+        conn = libvirt.open(configuration.libvirt_uri)
+        hv = vl.LibvirtHypervisor(conn)
+        hv.init_storage_pool(configuration.storage_pool)
+
+    configuration = configuration if configuration is not None else Configuration()
```
Comment (goneri) [inline virt_lightning/api.py:489]: You can acutally write this:

`configuration = configuration or Configuration()`

**Review by goneri — APPROVED** [review]

**Comment by goneri** [comment]
A green CI :partying_face:

**Comment by goneri** [comment]
Thank you @abikouo

**Pull request merged**